src/distributions/mod.rs

```diff
@@ -181,7 +181,11 @@ pub use self::other::Alphanumeric;
 #[doc(inline)] pub use self::uniform::Uniform;
 pub use self::float::{OpenClosed01, Open01};
 pub use self::bernoulli::Bernoulli;
-#[cfg(feature="alloc")] pub use self::weighted::{WeightedIndex, WeightedError};
+#[cfg(feature = "alloc")]
+pub use self::weighted::{
+    AliasMethodWeight, AliasMethodWeightedIndex, AliasMethodWeightedIndexError, WeightedError,
+    WeightedIndex,
```

Comment:
Alternatively we could make the module public and name these `weighted::{AliasMethodWeight, AliasMethod, Error, BinarySearch}`. Thoughts?

(obviously a breaking change)

Reply:
I think that would be better and more like it is done in the standard library  (`std::io::Error`). Again, let me know if you want to do that.

Reply:
But maybe the module must be renamed because with the proposed naming the 'index' part is lost.

Reply:
I'm not too sure we need to keep `Index` in the name anyway; it's the only type of weighted sampling we have. It's not the best idea to break this stuff *again*, but still better to get it right than leave a mess IMO.

But lets wait for @vks to comment.

Reply:
As I mentioned above, I think we might get rid of our current implementation and only have the alias method. Then there wouldn't be naming issues.

Alternatively, if we decide to keep both, I would prefer to drop the common `AliasMethod` prefix and instead have an `alias_method` module. This would make it much easier for users to switch between the two implementations.

Reply:
The current benchmarks show the alias method to always be in the lead or only slightly behind, *however* if you move the set-up time into the measurement loop, then the binary-search method can be significantly faster (three times faster on the large_set bench with 1000 samples; nine times faster with 100 samples, and a little faster on the smaller sets).

Memory usage will be a little higher with the Alias method due to the extra `Vec<usize>`; mostly this is unimportant I think (unless memory constrained and having a large set of weights in a small type).

The Alias method has some extra requirements on the type, notably `Copy`. Should we use `Clone` instead?

---

I think there is room for both implementations, though the current presentation and documentation is not ideal. So what do you think about the following structure?
```rust
distributions::{
    weighted::{
        alias::{WeightedIndex, Weight},
        WeightedIndex, Error
    },
}
```

Reply:
> The Alias method has some extra requirements on the type, notably Copy. Should we use Clone instead?

I missed that. In that case we should probably have both, and working with `Clone` would be nicer.

> So what do you think about the following structure?

This is what I would suggest as well.

Reply:
Good. @zroug would you make these changes please?

Reply:
Yes, I will make these changes.

Reply:
I did these changes in 5b2934122f04e1ff41024b3c0597ace711d21f08 but I used `alias_method` instead of `alias` as module name. Just `alias` didn't sound right to me and the word alias has a much broader meaning. Are you okay with this?

I wasn't sure if I should keep the reexports for `WeightedIndex`. I have kept them.